src/BenchmarksApps/SslStream/Server/Program.cs

```diff
@@ -0,0 +1,177 @@
+    }
+
+    static async Task Run(ServerOptions options)
+    {
+        var sslOptions = CreateSslServerAuthenticationOptions(options);
+        using var sock = new Socket(SocketType.Stream, ProtocolType.Tcp);
+        sock.Bind(new IPEndPoint(IPAddress.IPv6Any, options.Port));
+        sock.Listen();
+
+        Log($"Listening on {sock.LocalEndPoint}.");
+
+        CancellationTokenSource ctrlC = new CancellationTokenSource();
+        Console.CancelKeyPress += (s, e) =>
+        {
+            Log("Shutting down...");
+            e.Cancel = true;
+            ctrlC.Cancel();
+        };
+
+        try
+        {
+            while (!ctrlC.IsCancellationRequested)
+            {
+                var clientSock = await sock.AcceptAsync(ctrlC.Token).ConfigureAwait(false);
+                _ = Task.Run(() => ProcessClient(clientSock, options, sslOptions, ctrlC.Token));
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            // Expected, just return
+        }
+
+        Log("Exitting...");
+    }
+
+    static async Task ReadWriteScenario(SslStream stream, ServerOptions options, CancellationToken cancellationToken)
+    {
+        static async Task WritingTask(SslStream stream, int bufferSize, CancellationToken cancellationToken)
+        {
+            if (bufferSize == 0)
+            {
+                return;
+            }
+
+            var sendBuffer = new byte[bufferSize];
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    await stream.WriteAsync(sendBuffer, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected, just return
+            }
+        }
+
+        byte[] recvBuffer = new byte[options.ReceiveBufferSize];
+
+        CancellationTokenSource cts = new CancellationTokenSource();
+        using var reg = cancellationToken.Register(() => cts.Cancel());
+
+        var writeTask = Task.Run(() => WritingTask(stream, options.SendBufferSize, cts.Token));
+
+        var recvBufer = new byte[options.ReceiveBufferSize];
+
+        while (true)
+        {
+            if (await stream.ReadAsync(recvBufer, cts.Token).ConfigureAwait(false) == 0)
+            {
+                cts.Cancel();
+                break;
+            }
+        }
+
+        await writeTask.ConfigureAwait(false);
+    }
+
+    static async Task ProcessClient(Socket socket, ServerOptions options, SslServerAuthenticationOptions sslOptions, CancellationToken cancellationToken)
+    {
+        try
+        {
+            using var stream = await EstablishSslStreamAsync(socket, sslOptions, cancellationToken).ConfigureAwait(false);
+
+            if (stream.NegotiatedApplicationProtocol.Protocol.Span.SequenceEqual("Handshake"u8))
+            {
+                // done everything
+                return;
```

Comment:
Do I understand correctly that disposing the SslStream will also close the socket?

Reply:
Yes, the SslStream is set to dispose the inner NetworkStream which in turn is set to own the Socket. see `EstablishSslStreamAsync`